File: library/src/main/java/name/zeno/android/util/IOUtils.java
package name.zeno.android.util;

import android.content.Context;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;

import io.reactivex.Observable;


/**
 * @author 陈治谋 (513500085@qq.com)
 * @since 2016/10/13.
 */
public class IOUtils
{
  public static Observable<String> readString(Context context, String file)
  {
    return Observable.create(subscriber -> {
      InputStream is = null;
      try {
        is = context.getAssets().open(file);
      } catch (IOException e) {
        subscriber.onError(e);
      }

      if (is == null) {
        subscriber.onError(new IllegalArgumentException("file not found"));
        return;
      }

      try {
        subscriber.onNext(readString(is));
        subscriber.onComplete();
        is.close();
      } catch (IOException e) {
        subscriber.onError(e);
      }
    });
  }

  /**
   * 将InputStream转换成某种字符编码的String
   */
  public static String readString(InputStream is) throws IOException
  {
    ByteArrayOutputStream baos = new ByteArrayOutputStream();

    int i;
    while ((i = is.read()) != -1) {
      baos.write(i);
    }
    return baos.toString();
  }
}

File: library/src/main/java/name/zeno/android/util/IOUtils.kt
package name.zeno.android.util

import android.content.Context
import io.reactivex.Emitter

import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream

import io.reactivex.Observable


/**
 * @author 陈治谋 (513500085@qq.com)
 * @since 2016/10/13.
 */
object IOUtils {
  fun readString(context: Context, file: String): Observable<String> {
    return Observable.create(souse@ { subscriber: Emitter<String> ->
      var inputStream: InputStream? = null
      try {
        inputStream = context.assets.open(file)
      } catch (e: IOException) {
        subscriber.onError(e)
      }

      if (inputStream == null) {
        subscriber.onError(IllegalArgumentException("file not found"))
        return@souse
      }

      try {
        subscriber.onNext(readString(inputStream))
        subscriber.onComplete()
        inputStream.close()
      } catch (e: IOException) {
        subscriber.onError(e)
      }
    })
  }

  /** # 将InputStream转换成某种字符编码的String */
  @Throws(IOException::class)
  fun readString(inputStream: InputStream): String {
    val oStream = ByteArrayOutputStream()

    var i: Int
    do {
      i = inputStream.read()
      if (i != -1) oStream.write(i)
    } while (i != -1)
    return oStream.toString()
  }
}
